raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -565,20 +583,59 @@ private FetchQuorumRecordsResponseData buildEmptyFetchQuorumRecordsResponse(
      * - {@link Errors#INVALID_REQUEST} if the request epoch is larger than the leader's current epoch
      *     or if either the fetch offset or the last fetched epoch is invalid
      */
-    private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
+    private CompletableFuture<FetchQuorumRecordsResponseData> handleFetchQuorumRecordsRequest(
         RaftRequest.Inbound requestMetadata
     ) throws IOException {
         FetchQuorumRecordsRequestData request = (FetchQuorumRecordsRequestData) requestMetadata.data;
+        FetchQuorumRecordsResponseData response = tryCompleteFetchQuorumRecordsRequest(request);
+        if (response.errorCode() != Errors.NONE.code()
+            || response.records().hasRemaining()
+            || request.maxWaitTimeMs() == 0) {
+            return completedFuture(response);
+        }
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        purgatory.await(future, request.maxWaitTimeMs());
+
+        return future.handle((nil, exception) -> {
+            if (exception != null) {
+                Throwable cause = exception instanceof ExecutionException ?
+                    exception.getCause() : exception;
+
+                Errors error = Errors.forException(cause);
+                if (error != Errors.REQUEST_TIMED_OUT) {
+                    return buildEmptyFetchQuorumRecordsResponse(error, OptionalLong.empty());
+                }
+            }
+
+            try {
+                return tryCompleteFetchQuorumRecordsRequest(request);
+            } catch (Exception e) {
```

Comment:
This callback could be triggered either from the expiration reaper or in the context of handling another request as we update log end offset or become a follower. In the first case, the exception would just get lost; in the second case, it might actually cause a failure to update state as expected. Note also that if we propagated the exception, then we would fail to respond to the request.

Reply:
So the deal here is to avoid failing the reaper thread and make sure we could complete the response?

Reply:
That is part of it. As I said above, we may also have an exception raised during completion, which could occur when handling a fetch from a voter (for example).